hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java

```diff
@@ -125,118 +120,141 @@ public void addDuplicationStrategy(DuplicationStrategy strategy) {
+			this.listeners = (T[]) Array.newInstance( eventType.baseListenerInterface(), 1 );
+			this.listeners[0] = listener;
 		}
-	}
+		else {
+			final int size = this.listeners.length;
 
-	@Override
-	public void prependListener(T listener) {
-		internalPreprendListener( listener );
-		checkForArrayRefresh();
-	}
+			//noinspection unchecked
+			final T[] newCopy = (T[]) Array.newInstance( eventType.baseListenerInterface(), size+1 );
+
+			// put the new one first
+			newCopy[0] = listener;
 
-	private void internalPreprendListener(T listener) {
-		if ( listenerShouldGetAdded( listener ) ) {
-			internalPrepend( listener );
+			// and copy the rest after it
+			System.arraycopy( this.listeners, 0, newCopy, 1, size );
+
+			this.listeners = newCopy;
 		}
 	}
 
-	private boolean listenerShouldGetAdded(T listener) {
-		final List<T> ts = listenersAsList;
-		if ( ts == null ) {
-			listenersAsList = new ArrayList<>();
-			return true;
-			// no need to do de-dup checks
+	private void handleListenerAddition(T listener, Consumer<T> additionHandler) {
+		if ( listeners == null ) {
+			additionHandler.accept( listener );
+			return;
 		}
 
-		boolean doAdd = true;
-		strategy_loop: for ( DuplicationStrategy strategy : duplicationStrategies ) {
-			final ListIterator<T> itr = ts.listIterator();
-			while ( itr.hasNext() ) {
-				final T existingListener = itr.next();
+		final T[] localListenersRef = this.listeners;
+
+		for ( DuplicationStrategy strategy : duplicationStrategies ) {
+
+			// for each strategy, see if the strategy indicates that any of the existing
+			//		listeners match the listener being added.  If so, we want to apply that
+			//		strategy's action.  Control it returned immediately after applying the action
+			//		on match - meaning no further strategies are checked...
+
+			for ( int i = 0; i < localListenersRef.length; i++ ) {
+				final T existingListener = localListenersRef[i];
+				log.debugf(
+						"Checking incoming listener [`%s`] for match against existing listener [`%s`]",
+						listener,
+						existingListener
+				);
+
 				if ( strategy.areMatch( listener,  existingListener ) ) {
+					log.debugf( "Found listener match between `%s` and `%s`", listener, existingListener );
+
 					switch ( strategy.getAction() ) {
-						// todo : add debug logging of what happens here...
 						case ERROR: {
 							throw new EventListenerRegistrationException( "Duplicate event listener found" );
 						}
 						case KEEP_ORIGINAL: {
-							doAdd = false;
-							break strategy_loop;
+							log.debugf( "Skipping listener registration (%s) : `%s`", strategy.getAction(), listener );
+							return;
 						}
 						case REPLACE_ORIGINAL: {
-							checkAgainstBaseInterface( listener );
-							performInjections( listener );
-							itr.set( listener );
-							doAdd = false;
-							break strategy_loop;
+							log.debugf( "Replacing listener registration (%s) : `%s` -> ", strategy.getAction(), existingListener, listener );
```

Comment:
Is there a `%s` missing after ` -> ` above?

Reply:
Yes :(  @DavideD since you started a new branch/PR can you just fix this there...

Reply:
sure